contracts/src/v0.8/shared/call/CallWithExactGasZKSync.sol

```diff
@@ -0,0 +1,129 @@
+  /// that if a subcall is not trusted, it can consume lots of pubdata in the process. This function ensures that
+  /// no more than  `_maxTotalGas` will be allowed to be spent by the call. To be sure, this function uses some margin
+  /// (`CALL_ENTRY_OVERHEAD` + `CALL_RETURN_OVERHEAD`) to ensure that the call will not exceed the limit, so it may
+  /// actually spend a bit less than `_maxTotalGas` in the end.
+  /// @dev The entire `gas` passed to this function could be used, regardless
+  /// of the `_maxTotalGas` parameter. In other words, `max(gas(), _maxTotalGas)` is the maximum amount of gas that can be spent by this function.
+  /// @dev The function relays the `returndata` returned by the callee. In case the `callee` reverts, it reverts with the same error.
+  /// @param _to The address of the contract to call.
+  /// @param _maxTotalGas the maximum amount of gas that can be spent by the call.
+  /// @param _data The calldata for the call.
+  /// @param _maxReturnBytes the maximum amount of bytes that can be returned by the call.
+  /// @return success whether the call succeeded
+  /// @return gasUsed the gas used by the external call. Does not include the overhead of this function.
+  /// @return retData the return data from the call, capped at maxReturnBytes bytes
+  function _callWithExactGasSafeReturnData(
+    address _to,
+    uint256 _maxTotalGas,
+    bytes calldata _data,
+    uint16 _maxReturnBytes
+  ) external returns (bool, uint256, bytes memory) {
+    // At the start of the execution we deduce how much gas be spent on things that will be
+    // paid for later on by the transaction.
+    // The `expectedForCompute` variable is an upper bound of how much this contract can spend on compute and
+    // MUST be higher or equal to the `gas` passed into the call.
+    uint256 expectedForCompute = gasleft() + CALL_ENTRY_OVERHEAD;
+
+    // We expect that the `_maxTotalGas` at least includes the `gas` required for the call.
+    // This require is more of a safety protection for the users that call this function with incorrect parameters.
+    //
+    // Ultimately, the entire `gas` sent to this call can be spent on compute regardless of the `_maxTotalGas` parameter.
+    if (_maxTotalGas < gasleft()) {
+      revert GasLimitTooLow();
+    }
+
+    // This is the amount of gas that can be spent *exclusively* on pubdata in addition to the `gas` provided to this function.
+    uint256 pubdataAllowance = _maxTotalGas > expectedForCompute ? _maxTotalGas - expectedForCompute : 0;
+
+    uint256 pubdataPublishedBefore = SYSTEM_CONTEXT_CONTRACT.getCurrentPubdataSpent();
+
+    // We never permit system contract calls.
+    // If the call fails, the `EfficientCall.call` will propagate the revert.
+    // Since the revert is propagated, the pubdata published wouldn't change and so no
+    // other checks are needed.
+    // solhint-disable-next-line avoid-low-level-calls
+    bytes memory returnData = EfficientCall.call({
```

Comment:
i believe we need to limit the bytes from here?
bc if the call returns a long bytes, it still will be copied to `returnData`

in fact, i dug into the code behind `EfficientCall` and eventually track it down [here](https://github.com/matter-labs/era-contracts/blob/main/system-contracts/contracts/libraries/EfficientCall.sol#L229-L243)

i think we need to change the above section to ensure the entire solution is working just like the library [here](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/shared/call/CallWithExactGas.sol#L114-L125)
bc right after a native `call`, it check return data and never copies the full size return data if it's too big.

Reply:
for automation, we are fine with ignoring all the return data and only have `pubdataGas` in the return.

Reply:
cool great!

Reply:
Fixed, thanks!